aws/resource_aws_workspace.go

```diff
@@ -0,0 +1,405 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/workspaces"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags"
+)
+
+func resourceAwsWorkspace() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsWorkspaceCreate,
+		Read:   resourceAwsWorkspaceRead,
+		Update: resourceAwsWorkspaceUpdate,
+		Delete: resourceAwsWorkspaceDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+		Schema: map[string]*schema.Schema{
+			"bundle_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"directory_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"ip_address": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"computer_name": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"state": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"root_volume_encryption_enabled": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				ForceNew: true,
+				Default:  false,
+			},
+			"user_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"user_volume_encryption_enabled": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				ForceNew: true,
+				Default:  false,
+			},
+			"volume_encryption_key": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"workspace_properties": {
```

Comment:
In this case, I can't specify workspaces properties as a block construct, right? I find the list declaration for key-value properties a little bit weird.

Reply:
Ah, I guess I'm wrong – `aws_worskpaces_directory.self_service_permission` attribute is `List`, but it could be declared with the block in Terraform:
```
resource "aws_workspaces_directory" "main" {
  ...
  self_service_permissions {
    change_compute_type = false
    increase_volume_size = true
    rebuild_workspace = true
    restart_workspace = false
    switch_running_mode = true
  }
```